### PR TITLE
Improved nullable '=='/'!=' analysis

### LINKED ISSUE #41107: Incorrect nullable analysis of `x?.y.Extension() == true` pattern
image

``` csharp

using System.Diagnostics.CodeAnalysis;

#nullable enable

public class C 
{
    public C? field;
    
    static public void M(C c) 
    {
       if (c.field?.field.IsKind() == true)
       {
          c.field.field.ToString(); // warning CS8602: Dereference of a possibly null reference.
       }
    }
    
}

public static class Extension
{
    public static bool IsKind([NotNullWhen(true)] this C? c)
    {
        throw null!;
    }
}
```

[sharplab repro](https://sharplab.io/#v2:EYLgxg9gTgpgtADwGwBYA+BYAUAAQAwAEOAjAHQAiAlgIYDmAdhAM4AulYTpAwhACYwBBetQA2ATyaUmAbmzYAxPQCuIkdWAiYBGMI0w5uAMxEATAS4FsAb2wE7RY1wD8BAGaUYI3rKz3Lv+xIkByIUAgBZAAoLMABKfz8bAL8CSlcCSLBSd09eJ2yPL1IASSYAaUp6Xkj4gF5aghYoJRhY2xSCJI6/LJyivt5SABUIAGUmytoa6QIAelmCAHdqKHpJ81GADiQ8ExACchhYVyOdMC0IdOoCAAdmSQ0xAmVVAmPT+nPSdpSAXx+CP9kv4gdgcMYgqYCABRBAsHSSCD0awA8FEYjBYAQCAiAilCpVSIAbQAchAWCSVCIAOoACx0kSaLViAF1GrSpOYXHEAV1uixaVAIItnlSAIQ+PxAoFAA===)

### PR DESCRIPTION
**Pull request by RikkiGibson**
Closes #41107

Test plan: #51463

Corresponding definite assignment PR: #52425

Specification (for definite assignment, but many aspects have an analogous fit in nullable analysis): https://github.com/dotnet/csharplang/blob/main/proposals/improved-definite-assignment.md#-relational-equality-operator-expressions

### CODE REVIEW

File: src/Compilers/CSharp/Portable/FlowAnalysis/NullableWalker.cs [hunk 1]
```diff
@@ -3805,6 +3893,37 @@ private void AfterLeftChildHasBeenVisited(BoundExpression leftOperand, Conversio
             var inferredResult = InferResultNullability(binary.OperatorKind, method, binary.Type, leftType, rightType);
             SetResult(binary, inferredResult, inferredResult.ToTypeWithAnnotations(compilation));
 
+            TypeSymbol? getTypeIfContainingType(TypeSymbol baseType, TypeSymbol? derivedType)
```
Comment (RikkiGibson) [inline src/Compilers/CSharp/Portable/FlowAnalysis/NullableWalker.cs:3896]: This is just a move.

File: src/Compilers/CSharp/Portable/FlowAnalysis/NullableWalker.cs [hunk 2]
```diff
@@ -3621,13 +3621,61 @@ private TypeWithState InferResultNullability(BinaryOperatorKind operatorKind, Me
...
+                var leftType = ResultType;
+                var (rightOperand, rightConversion) = RemoveConversion(binary.Right, includeExplicitConversions: false);
+
+                // Consider the following two scenarios:
+                // `a?.b(x = new object()) == c.d(x = null)` // `x` is maybe-null after expression
+                // `a?.b(x = null) == c.d(x = new object())` // `x` is not-null after expression
+
+                // In order to properly integrate state changes from the RHS into the LHS's "state when not null", we can't simply Meet or Join the state.
+                // We have to visit it twice. Once without diagnostics using the LHS's "state when not null", and again with the "state when maybe null" for diagnostics and public API.
+                var oldDisableDiagnostics = _disableDiagnostics;
```
Comment (RikkiGibson) [inline src/Compilers/CSharp/Portable/FlowAnalysis/NullableWalker.cs:3644]: This part is fairly gnarly. I wasn't able to find a complete solution that didn't involve double-visiting the right operand. Because the operator chain is left-associative, we think that re-visiting the right in this scenario is not that expensive.

If this is found to be too ugly of a solution, we could also consider adding a requirement that the "other" operand is a constant value in order to learn from a conditional access.

**Comment by RikkiGibson** [comment]
@333fred @jcouv Please take a look when you have the chance.

**Comment by RikkiGibson** [comment]
@333fred @jcouv Please review

**Comment by jcouv** [comment]
Looking

File: src/Compilers/CSharp/Portable/FlowAnalysis/AbstractFlowPass.cs [hunk 3]
```diff
@@ -2713,30 +2713,30 @@ private bool TryVisitConditionalAccess(BoundExpression node, [NotNullWhen(true)]
...
+            if (method is null)
             {
-                return true;
+                // This can happen when the user-defined conversion to use is ambiguous.
+                // For example in `SyntaxBinderTests.TestNullCoalesceWithInvalidUserDefinedConversions_03`
+                return false;
             }
 
-            return false;
+            Debug.Assert(method.ParameterCount is 1);
```
Comment (333fred) [inline src/Compilers/CSharp/Portable/FlowAnalysis/AbstractFlowPass.cs:2737]: Is this actually true for all error scenarios?

Comment (333fred) [reply src/Compilers/CSharp/Portable/FlowAnalysis/AbstractFlowPass.cs:2737]: And if it isn't, what would we even do about it?

File: src/Compilers/CSharp/Portable/FlowAnalysis/NullableWalker.cs [hunk 4]
```diff
@@ -3645,74 +3693,114 @@ protected override void VisitBinaryOperatorChildren(ArrayBuilder<BoundBinaryOper
                 binary = stack.Pop();
             }
 
-            // learn from true/false constant
-            bool learnFromBooleanConstantTest()
+            static bool isEqualsOrNotEquals(BoundBinaryOperator binary)
+                => binary.OperatorKind.Operator() is BinaryOperatorKind.Equal or BinaryOperatorKind.NotEqual;
```
Comment (333fred) [inline src/Compilers/CSharp/Portable/FlowAnalysis/NullableWalker.cs:3697]: Honestly feels like these should be inlined. I was expecting the logic to be more complicated because you broke it out into a separate function, but it's actually a pretty simple check.

File: src/Compilers/CSharp/Portable/FlowAnalysis/AbstractFlowPass.cs (hunk 3 above)
Comment (RikkiGibson) [reply src/Compilers/CSharp/Portable/FlowAnalysis/AbstractFlowPass.cs:2737]: > Is this actually true for all error scenarios?

I actually don't know. I'll add a test.

> And if it isn't, what would we even do about it?

I think we're in "degrade gracefully" mode at that point. I would be fine with just using the first parameter if one exists, and otherwise saying that the conversion can't propagate "state when not null".

File: src/Compilers/CSharp/Portable/FlowAnalysis/NullableWalker.cs [hunk 5]
```diff
@@ -3645,74 +3693,114 @@ protected override void VisitBinaryOperatorChildren(ArrayBuilder<BoundBinaryOper
...
+                    stateWhenNotNull = conditionalStateWhenNotNull.State;
                 }
-
-                if (!leftConversion.IsIdentity)
+                else if (isEquals(binary) && otherOperand.ConstantValue is { IsBoolean: true, BooleanValue: var boolValue })
                 {
-                    return false;
+                    // can preserve conditional state from `.TryGetValue` in `dict?.TryGetValue(key, out value) == true`,
+                    // but not in `dict?.TryGetValue(key, out value) != false`
+                    stateWhenNotNull = boolValue ? conditionalStateWhenNotNull.StateWhenTrue : conditionalStateWhenNotNull.StateWhenFalse;
```
Comment (333fred) [inline src/Compilers/CSharp/Portable/FlowAnalysis/NullableWalker.cs:3719]: But what about `is true or false` (just to throw more curveballs at you :P).

Comment (RikkiGibson) [reply src/Compilers/CSharp/Portable/FlowAnalysis/NullableWalker.cs:3719]: Coming in #53311 :)

Comment (333fred) [inline src/Compilers/CSharp/Portable/FlowAnalysis/NullableWalker.cs:3719]: Actual "but what about" question: what about `!= null`?

Comment (RikkiGibson) [reply src/Compilers/CSharp/Portable/FlowAnalysis/NullableWalker.cs:3719]: If the conditional access is accessing a member with postconditions like `[NotNullWhen(...)]` then we unsplit the conditional "state when not null" in the `else` block on line 3721, then use it as either the `StateWhenTrue` or `StateWhenFalse` as applicable for after the comparison.

File: src/Compilers/CSharp/Test/Semantic/Semantics/NullableReferenceTypesTests.cs [hunk 6]
```diff
@@ -28932,16 +28932,7 @@ public static bool IsKind([NotNullWhen(true)] this C? c)
...
-            // If we wanted to analyze this properly, we would probably need a WhenTrue/WhenFalse/WhenNull split.
-            c.VerifyDiagnostics(
-                // (12,13): warning CS8602: Dereference of a possibly null reference.
-                //             c.field.field.ToString(); // 1
-                Diagnostic(ErrorCode.WRN_NullReferenceReceiver, "c.field.field").WithLocation(12, 13),
-                // (20,13): warning CS8602: Dereference of a possibly null reference.
-                //             c.field.field.ToString(); // 2
-                Diagnostic(ErrorCode.WRN_NullReferenceReceiver, "c.field.field").WithLocation(20, 13)
-                );
+            c.VerifyDiagnostics();
```
Comment (jcouv) [inline src/Compilers/CSharp/Test/Semantic/Semantics/NullableReferenceTypesTests.cs:28935]: Yay! :-)

File: src/Compilers/CSharp/Test/Semantic/Semantics/NullableReferenceTypesTests.cs [hunk 7]
```diff
@@ -47139,6 +47130,1576 @@ void M5(object? x)
...
+            CreateNullableCompilation(source).VerifyDiagnostics(
+                // (9,15): warning CS8602: Dereference of a possibly null reference.
+                //             : x.ToString(); // 1
+                Diagnostic(ErrorCode.WRN_NullReferenceReceiver, "x").WithLocation(9, 15)
+                );
+        }
+
+        /// <remarks>Ported from <see cref="FlowTests.EqualsCondAccess_14"/>.</remarks>
+        [Fact]
+        public void EqualsCondAccess_14()
```
Comment (jcouv) [inline src/Compilers/CSharp/Test/Semantic/Semantics/NullableReferenceTypesTests.cs:48391]: If not already covered, consider adding a similar test where the body starts with `if (i is null) throw ...;`

File: src/Compilers/CSharp/Test/Semantic/Semantics/NullableReferenceTypesTests.cs [hunk 8]
```diff
@@ -47139,6 +47130,1576 @@ void M5(object? x)
...
+{
+    void M(C? c, object? x, object? y)
+    {
+        _ = (c?.Equals(x = 0), c?.Equals(y = 0)) == (true, true)
+            ? x.ToString() // 1
+            : y.ToString(); // 2
+    }
+}
+";
+            // This could be made to work (i.e. removing diagnostic 1) but isn't a high priority scenario.
```
Comment (jcouv) [inline src/Compilers/CSharp/Test/Semantic/Semantics/NullableReferenceTypesTests.cs:48472]: >but isn't a high priority scenario. [](http://example.com/codeflow?start=71&length=35)

nit: I'd remove this comment. Just say issue is tracked

**Comment by jcouv** [comment]
Short meeting notes (5/18/2021): Aleksey, Fred, Rikki and I thoroughly explored options, including an approach tracking changed slots. We're thinking that the double-visit approach is still likely the best trade-off (usability/complexity/performance), and we'll do some measurement to validate performance behavior.

File: src/Compilers/CSharp/Portable/FlowAnalysis/NullableWalker.cs [hunk 9]
```diff
@@ -3645,74 +3693,114 @@ protected override void VisitBinaryOperatorChildren(ArrayBuilder<BoundBinaryOper
...
+            static bool isEquals(BoundBinaryOperator binary)
+                => binary.OperatorKind.Operator() == BinaryOperatorKind.Equal;
+
+            static bool isKnownNullOrNotNull(BoundExpression expr, TypeWithState resultType)
+            {
+                return resultType.State.IsNotNull()
+                    || expr.ConstantValue is object;
+            }
+
+            LocalState getSingleState(BoundExpression otherOperand, PossiblyConditionalState conditionalStateWhenNotNull)
```
Comment (jcouv) [inline src/Compilers/CSharp/Portable/FlowAnalysis/NullableWalker.cs:3708]: >getSingleState [](http://example.com/codeflow?start=23&length=14)

nit: `getUnconditionalState`?

File: src/Compilers/CSharp/Portable/FlowAnalysis/NullableWalker.cs [hunk 10]
```diff
@@ -3621,13 +3621,61 @@ private TypeWithState InferResultNullability(BinaryOperatorKind operatorKind, Me
...
+            {
+                Debug.Assert(!IsConditionalState);
+                var leftType = ResultType;
+                var (rightOperand, rightConversion) = RemoveConversion(binary.Right, includeExplicitConversions: false);
+
+                // Consider the following two scenarios:
+                // `a?.b(x = new object()) == c.d(x = null)` // `x` is maybe-null after expression
+                // `a?.b(x = null) == c.d(x = new object())` // `x` is not-null after expression
+
+                // In order to properly integrate state changes from the RHS into the LHS's "state when not null", we can't simply Meet or Join the state.
```
Comment (jcouv) [inline src/Compilers/CSharp/Portable/FlowAnalysis/NullableWalker.cs:3642]: nit: Consider refreshing the comment in light of our discussion today. Maybe something like:

// If the comparison is amenable:
// We take the when-not-null state from visiting the LHS and use that as starting point to visit the RHS ignoring diagnostics. The resulting state is the when-true state resulting from the comparison.
// We take the state (ie. worst case state) from visiting the LHS and use that as starting point to visit the RHS for diagnostics and recording nullabilities for public APIs. The resulting state is the when-false state resulting from the comparison.

Comment (RikkiGibson) [reply src/Compilers/CSharp/Portable/FlowAnalysis/NullableWalker.cs:3642]: > The resulting state is the when-true state

Whether the resulting state is for when-true or when-false of the containing comparison depends on the operator and the "other" operand in use, roughly according to the following

| Operator | Other operand | StateWhenNotNull moves to |
|----------|---------------|---------------------------|
| ==       | non-null      | StateWhenTrue             |
| ==       | constant null | StateWhenFalse            |
| !=       | non-null      | StateWhenFalse            |
| !=       | constant null | StateWhenTrue             |

File: src/Compilers/CSharp/Portable/FlowAnalysis/NullableWalker.cs [hunk 11]
```diff
@@ -3621,13 +3621,61 @@ private TypeWithState InferResultNullability(BinaryOperatorKind operatorKind, Me
...
+                VisitRvalue(rightOperand);
+                var stateWhenNotNull = this.State;
+
+                _disableDiagnostics = oldDisableDiagnostics;
+
+                // Now visit the right side for public API and diagnostics using the worst-case state from the LHS.
+                SetState(stateBeforeRight);
+                var rightType = VisitRvalueWithState(rightOperand);
+                ReinferBinaryOperatorAndSetResult(leftOperand, leftConversion, leftType, rightOperand, rightConversion, rightType, binary);
+                if (isKnownNullOrNotNull(rightOperand, rightType))
```
Comment (jcouv) [inline src/Compilers/CSharp/Portable/FlowAnalysis/NullableWalker.cs:3658]: I seems that when this condition is false, we'll have done the first visit (based on when-not-null state) for nothing. 
I suspect we can re-organize the code to do the second visit (based on worst-case state) first, then decide if we need to do the first visit (based on when-not-null state).

**Review by jcouv — COMMENTED** [review]
Done with review pass (iteration 1)

File: src/Compilers/CSharp/Portable/FlowAnalysis/NullableWalker.cs (hunk 11 above)
Comment (RikkiGibson) [reply src/Compilers/CSharp/Portable/FlowAnalysis/NullableWalker.cs:3658]: I made a point to do the worst-case visit last because I am concerned that subsequent visits will overwrite values from previous visits that get used for public API. It might be good to add some public API tests for this PR, for example to check the flow state for `x` on the RHS of `a(x = null)?.b(x = 1) == M(x)`, for example. (we would expect a maybe-null flow state).

Comment (RikkiGibson) [reply src/Compilers/CSharp/Portable/FlowAnalysis/NullableWalker.cs:3658]: /cc @333fred in case you have thoughts on this specific part of the implementation or on what should be covered in any new public API tests.

File: src/Compilers/CSharp/Test/Semantic/Semantics/NullableReferenceTypesTests.cs [hunk 12]
```diff
@@ -47139,6 +47130,1576 @@ void M5(object? x)
...
+                Diagnostic(ErrorCode.WRN_NullReferenceReceiver, "y").WithLocation(30, 30),
+                // (31,30): warning CS8602: Dereference of a possibly null reference.
+                //             : x.ToString() + y.ToString(); // 8
+                Diagnostic(ErrorCode.WRN_NullReferenceReceiver, "y").WithLocation(31, 30)
+                );
+        }
+
+        /// <remarks>Ported from <see cref="FlowTests.EqualsCondAccess_04"/>.</remarks>
+        [Fact]
+        public void EqualsCondAccess_04_03()
```
Comment (RikkiGibson) [inline src/Compilers/CSharp/Test/Semantic/Semantics/NullableReferenceTypesTests.cs:47906]: I think we should also have a test like `if (a?.b(out x) == M0(y = x)) y.ToString();`

**Comment by RikkiGibson** [comment]
Have briefly investigated the performance with various complex right-hand sides.

<details>
    <summary>Benchmark source</summary>

```cs
[SimpleJob(RuntimeMoniker.NetCoreApp50)]
public class ComplexCondAccessEquals
{
    private MetadataReference[] coreRefs = null!;

    [GlobalSetup]
    public void Setup()
    {
        coreRefs = new[] { MetadataReference.CreateFromFile(typeof(object).Assembly.Location) };
    }

    [Params(10, 12, 15, 18)]
    public int Depth { get; set; }

    [Params("enable", "disable")]
    public string NullableContext { get; set; } = null!;

    [Benchmark]
    public void CondAccess_ComplexRightSide()
    {
        var source1 = @"
#nullable " + NullableContext + @"
object? x = null;
C? c = null;
if (
";
        var source2 = @"
c!.M(x))
{
}

class C
{
public bool M(object? obj) => false;
}
";
        var sourceBuilder = new StringBuilder();
        sourceBuilder.Append(source1);
        for (var i = 0; i < Depth; i++)
        {
            sourceBuilder.AppendLine($"    c?.M(x = {i}) == (");
        }

        sourceBuilder.Append("    ");
        for (var i = 0; i < Depth; i++)
        {
            sourceBuilder.Append(")");
        }

        sourceBuilder.Append(source2);

        var tree = SyntaxFactory.ParseSyntaxTree(sourceBuilder.ToString(), path: "bench.cs");
        var comp = CSharpCompilation.Create(
            "Benchmark",
            new[] { tree },
            coreRefs);
        var diags = comp.GetDiagnostics();
    }
}
```
</details>

BenchmarkDotNet=v0.12.1, OS=Windows 10.0.19042
AMD Ryzen 9 3900X, 1 CPU, 24 logical and 12 physical cores
.NET Core SDK=6.0.100-preview.3.21202.5
  [Host]        : .NET Core 5.0.4 (CoreCLR 5.0.421.11614, CoreFX 5.0.421.11614), X64 RyuJIT
  .NET Core 5.0 : .NET Core 5.0.4 (CoreCLR 5.0.421.11614, CoreFX 5.0.421.11614), X64 RyuJIT

Job=.NET Core 5.0  Runtime=.NET Core 5.0  

|                      Method | Depth | NullableContext |         Mean |      Error |     StdDev |       Median |
|---------------------------- |------ |---------------- |-------------:|-----------:|-----------:|-------------:|
| CondAccess_ComplexRightSide |    10 |         disable |     2.022 ms |  0.1183 ms |  0.3412 ms |     2.029 ms |
| CondAccess_ComplexRightSide |    10 |          enable |    15.250 ms |  0.3026 ms |  0.6448 ms |    14.961 ms |
| CondAccess_ComplexRightSide |    12 |         disable |     1.966 ms |  0.0393 ms |  0.0941 ms |     1.953 ms |
| CondAccess_ComplexRightSide |    12 |          enable |    51.099 ms |  0.4954 ms |  0.4391 ms |    51.060 ms |
| CondAccess_ComplexRightSide |    15 |         disable |     2.052 ms |  0.0739 ms |  0.2133 ms |     2.019 ms |
| CondAccess_ComplexRightSide |    15 |          enable |   398.303 ms |  6.2167 ms |  5.8151 ms |   396.745 ms |
| CondAccess_ComplexRightSide |    18 |         disable |     2.174 ms |  0.0672 ms |  0.1905 ms |     2.126 ms |
| CondAccess_ComplexRightSide |    18 |          enable | 3,171.458 ms | 43.5458 ms | 38.6022 ms | 3,171.348 ms |

It seems to me like the level of performance we're seeing here is adequate. It certainly starts to get bad as the depth increases, but that's approaching the point where we expect only machine-generated code to reach, and machine-generated code like this should probably disable the nullable context regardless.

/cc @jcouv @AlekseyTs @333fred

File: src/Compilers/CSharp/Portable/FlowAnalysis/NullableWalker.cs (hunk 10 above)
Comment (jcouv) [reply src/Compilers/CSharp/Portable/FlowAnalysis/NullableWalker.cs:3642]: I'm fine if the comment only explains `==`. The swapping for `!=` was not hard to follow in my opinion.

File: src/Compilers/CSharp/Portable/FlowAnalysis/AbstractFlowPass.cs (hunk 3 above)
Comment (RikkiGibson) [reply src/Compilers/CSharp/Portable/FlowAnalysis/AbstractFlowPass.cs:2737]: Fixing this method by imitating logic from VisitUserDefinedConversion.

File: src/Compilers/CSharp/Portable/FlowAnalysis/NullableWalker.cs (hunk 4 above)
Comment (RikkiGibson) [reply src/Compilers/CSharp/Portable/FlowAnalysis/NullableWalker.cs:3697]: I like having `isEquals` as its own method because several calls to it are operands of `==` and it feels a little cleaner than parenthesizing a nested `==`, for example. But I will go ahead and eliminate `isEqualsOrNotEquals`.

**Comment by RikkiGibson** [comment]
@jcouv @333fred I believe I have addressed all the comments. Please take another look when you get the chance.

File: src/Compilers/CSharp/Portable/FlowAnalysis/NullableWalker.cs (hunk 11 above)
Comment (333fred) [reply src/Compilers/CSharp/Portable/FlowAnalysis/NullableWalker.cs:3664]: A good test would be calling GetTypeInfo on an expression on the RHS that is not-null in the when-not-null case and null in the worst case, and making sure that the type's state is null.

File: src/Compilers/CSharp/Test/Semantic/Semantics/OperatorTests.cs [hunk 13]
```diff
@@ -4979,6 +4979,40 @@ struct A
...
+        public void TestNullCoalesceWithInvalidUserDefinedConversions_04()
+        {
+            var source = @"
+struct B
+{
+    static void Main()
+    {
+        A a2;
+        B? b2 = null;
+        var c2 = b2 ?? a2;
```
Comment (333fred) [inline src/Compilers/CSharp/Test/Semantic/Semantics/OperatorTests.cs:4993]: Consider making this just an assignment, rather than relying on the rules of `??` to probably do the thing you want.

File: src/Compilers/CSharp/Portable/FlowAnalysis/NullableWalker.cs (hunk 11 above)
Comment (333fred) [reply src/Compilers/CSharp/Portable/FlowAnalysis/NullableWalker.cs:3664]: I think the tests you added cover this sufficiently.

**Comment by 333fred** [comment]
> <?xml version="1.0" encoding="utf-8"?>

We have an existing benchmarks project at dotnet/performance. I don't think we should add this.

---
Refers to: src/Tools/CompilerBenchmarks/CompilerBenchmarks.csproj:1 in 4718b68. [](commit_id = 4718b68b4474b2d7de81b1a1efb24d10842a4a98, deletion_comment = False)

**Comment by 333fred** [comment]
Done review pass (commit 4)

**Comment by RikkiGibson** [comment]
> We have an existing benchmarks project at dotnet/performance. I don't think we should add this.

Perhaps the right thing is just to remove the new project and let the test case live on in OverloadResolutionPerfTests.cs.

File: src/Compilers/CSharp/Test/Semantic/Semantics/OperatorTests.cs (hunk 13 above)
Comment (RikkiGibson) [reply src/Compilers/CSharp/Test/Semantic/Semantics/OperatorTests.cs]: Ended up replacing this with a test in FlowTests which attempts more directly to get into a situation where `CanPropagateStateWhenNotNull` could be called with a user-defined conversion with an unexpected parameter count. The method is shared between definite assignment and nullable so I feel like the level of testing introduced in the latest commit is adequate.

**Comment by RikkiGibson** [comment]
@333fred @jcouv I believe I've addressed all your comments. Please let me know if you have any more.

File: src/Compilers/CSharp/Test/Semantic/Semantics/NullableReferenceTypesTests.cs (hunk 7 above)
Comment (RikkiGibson) [reply src/Compilers/CSharp/Test/Semantic/Semantics/NullableReferenceTypesTests.cs:48391]: Done

File: src/Compilers/CSharp/Test/Semantic/Semantics/NullableReferenceTypesTests.cs (hunk 8 above)
Comment (RikkiGibson) [reply src/Compilers/CSharp/Test/Semantic/Semantics/NullableReferenceTypesTests.cs:48472]: Done

File: src/Compilers/CSharp/Portable/FlowAnalysis/NullableWalker.cs (hunk 9 above)
Comment (RikkiGibson) [reply src/Compilers/CSharp/Portable/FlowAnalysis/NullableWalker.cs]: Have changed the name to try and improve clarity

File: src/Compilers/CSharp/Portable/FlowAnalysis/NullableWalker.cs (hunk 10 above)
Comment (RikkiGibson) [reply src/Compilers/CSharp/Portable/FlowAnalysis/NullableWalker.cs]: Have updated the comment.

**Review by 333fred — APPROVED** [review]
LGTM (commit 7)

**Review by jcouv — APPROVED** [review]
LGTM Thanks (iteration 7)
